crates/autopilot/src/database/onchain_order_events/metrics.rs

```diff
@@ -0,0 +1,139 @@
+use super::OnchainOrderData;
+use database::{
+    events::EventIndex, onchain_broadcasted_orders::OnchainOrderPlacement, orders::OrderClass,
+};
+use itertools::Itertools;
+use model::order::OrderUid;
+use shared::{
+    current_block::RangeInclusive,
+    event_handling::MAX_REORG_BLOCK_COUNT,
+    orderbook_metrics::{db_order_class_label, operation_label, OrderOperation},
+};
+use std::collections::{HashMap, HashSet};
+use strum::IntoEnumIterator;
+
+#[derive(prometheus_metric_storage::MetricStorage, Clone, Debug)]
+#[metric(subsystem = "onchain_orders")]
+pub struct Metrics {
+    /// Keeps track of errors in picking up onchain orders.
+    /// Note that an order might be created even if an error is encountered.
+    #[metric(labels("error_type"))]
+    onchain_order_errors: prometheus::IntCounterVec,
+
+    /// Counts the number of onchain orders that were created or cancelled.
+    /// It accounts for reorg, which means that its value might decrease.
+    /// It might provide incorrect data if a reorg happens during a restart of
+    /// the services.
+    #[metric(labels("kind", "operation"))]
+    orders: prometheus::IntGaugeVec,
+}
+
+impl Metrics {
+    pub fn get() -> &'static Self {
+        Self::instance(global_metrics::get_metric_storage_registry())
+            .expect("unexpected error getting metrics instance")
+    }
+
+    pub fn init(&self) {
+        for (class, op) in OrderClass::iter().cartesian_product(OrderOperation::iter()) {
+            self.change_orders_by(0, &class, &op);
+        }
+
+        // Note: `onchain_order_errors` isn't initialized because keeping track
+        // of all possible error labels would require more maintenance than
+        // it's worth it.
+    }
+
+    pub fn inc_onchain_order_errors(&self, error_label: &str) {
+        self.onchain_order_errors
+            .with_label_values(&[error_label])
+            .inc();
+    }
+
+    pub fn change_orders_by(&self, amount: i64, class: &OrderClass, op: &OrderOperation) {
+        let class = db_order_class_label(class);
+        let op = operation_label(op);
+        self.orders.with_label_values(&[class, op]).add(amount);
+    }
+}
+
+type OrderDetailsForMetrics = (u64, OrderClass, OrderOperation);
+pub type OrderDetailsToOrderUids = HashMap<OrderDetailsForMetrics, HashSet<OrderUid>>;
+#[derive(Debug, Default)]
+pub struct OrdersByBlockNumber(OrderDetailsToOrderUids);
+
+impl OrdersByBlockNumber {
```

Comment:
Do you think the following way would be less complex and easier to maintain? 
run the following  SQL query 
`select count(*), orderclass from onchain_orders oo inner join orders o o.uid = oo.uid group by orderclass`
before a reorg and after a reorg and then simply update the counter by the difference. I hope this would be simpler code with less maintenance.